php_phongo.c

```diff
@@ -136,28 +136,35 @@ zend_class_entry* phongo_exception_from_phongo_domain(php_phongo_error_domain_t
 }
 zend_class_entry* phongo_exception_from_mongoc_domain(uint32_t /* mongoc_error_domain_t */ domain, uint32_t /* mongoc_error_code_t */ code)
 {
-	switch (code) {
-		case MONGOC_ERROR_STREAM_SOCKET:
-		case MONGOC_ERROR_SERVER_SELECTION_FAILURE:
-			return php_phongo_connectiontimeoutexception_ce;
-		case MONGOC_ERROR_CLIENT_AUTHENTICATE:
-			return php_phongo_authenticationexception_ce;
-		case MONGOC_ERROR_COMMAND_INVALID_ARG:
-			return php_phongo_invalidargumentexception_ce;
+	if (domain == MONGOC_ERROR_CLIENT && code == MONGOC_ERROR_CLIENT_AUTHENTICATE) {
+		return php_phongo_authenticationexception_ce;
 	}
-	switch (domain) {
-		case MONGOC_ERROR_COMMAND:
-			return php_phongo_commandexception_ce;
-		case MONGOC_ERROR_SERVER:
-			if (code == 50) {
-				return php_phongo_executiontimeoutexception_ce;
-			}
-			return php_phongo_serverexception_ce;
-		case MONGOC_ERROR_STREAM:
-			return php_phongo_connectionexception_ce;
-		default:
-			return php_phongo_runtimeexception_ce;
+
+	if (domain == MONGOC_ERROR_COMMAND && code == MONGOC_ERROR_COMMAND_INVALID_ARG) {
+		return php_phongo_invalidargumentexception_ce;
 	}
+
+	if (domain == MONGOC_ERROR_SERVER) {
+		if (code == 50) {
+			return php_phongo_executiontimeoutexception_ce;
+		}
+
+		return php_phongo_serverexception_ce;
+	}
+
+	if (domain == MONGOC_ERROR_SERVER_SELECTION && code == MONGOC_ERROR_SERVER_SELECTION_FAILURE) {
+		return php_phongo_connectiontimeoutexception_ce;
+	}
+
+	if (domain == MONGOC_ERROR_STREAM) {
+		if (code == MONGOC_ERROR_STREAM_SOCKET) {
+			return php_phongo_connectiontimeoutexception_ce;
+		}
+
+		return php_phongo_connectionexception_ce;
+	}
+
+	return php_phongo_runtimeexception_ce;
```

Comment:
I opted to remove the `switch` entirely since we are frequently matching domains and codes together. Also, a single `default:` would not have been sufficient, since some of the domains would have needed to `break;` out of the switch entirely and require this final `return`.

Reply:
This looks good, but it seems to me that the codes are unique so is it not redundant to also check the domain for the cases where there is only one outcome, such as `domain == MONGOC_ERROR_COMMAND && code == MONGOC_ERROR_COMMAND_INVALID_ARG`?

Reply:
Since some codes may come from the server, it's theoretically possible that a server-side code might match one of libmongoc's constants. In that case, I wanted to be defensive so that we're only considering codes for the expected domain.